docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,295 @@
+# Flow Modules
+
+* Date: 2023-01-27
+* Author: Matt Durham @mattdurham
+* PR: [grafana/agent#2898](https://github.com/grafana/agent/pull/2898)
+* Status: Draft
+
+[Formatted Link for ease of user](https://github.com/grafana/agent/blob/rfc_modules/docs/rfcs/0007-flow-modules.md)
+
+## RFC Goals
+
+* Explain the use cases of modules
+* Explain what modules are
+* Go over possible syntax for modules
+* Go over pros and cons of modules
+
+## Summary
+
+One of the primary goals for the production usage of Agent Flow is parity with the static subsystem. One of the features of the static subsystem is [scraping service](https://github.com/grafana/agent/blob/main/docs/sources/configuration/scraping-service.md). Scraping service allows a user to run a fleet of agents and have thousands of scrape configurations distributed among running Agents. Through discussions within the Agent Team, we did not want to limit dynamically loading content to only scrape configs but allow most components to be loaded and used.
+
+During this time the Agent team saw a lot of potential in modules outside of scraping service. Packaging up sets of components for specific workflows, publishing common use cases and allowing better usage for internal customers in the Agent as a Service model.
+
+## Goals
+
+* Support single module loading via `module.single`
+* Support multiple module loading via `module.multiple`
+* Enable re-use of common patterns
+* Allow loading a module from a string
+* Allow modules to load other modules
+* Modules should be sandboxed except via arguments and exports
+
+### Enable re-use of common patterns
+
+Common functionality can be wrapped in a set of common components that form a module. These shared modules can then be used instead of reinventing use cases. 
+
+### Allow loading a module from a string
+
+Modules will not care about the source of a string. In the case of a `module.single` the module will take in one string of valid river configuration. In the case of `module.multiple` the module will take in a `map(string)` where the key uniquely denotes the module.
+
+### Allow modules to load other modules
+
+Modules will be able to load other modules, with reasonable safe guards. There will be a stack limit for the depth of sub-modules and modules will be unable to load themselves.
+
+### Modules should be sandboxed except via arguments and exports
+
+Modules cannot directly access children or parent modules except through predefined arguments and exports. 
+
+## Non Goals
+
+* Add additional capabilities to load strings
```

Comment:
"load strings" here feels ambiguous. I'm assuming you're talking about module sources? 

I'd like to suggest we introduce three new terms for this RFC to make things clearer:

* Module source: a Flow config file which will be used as a module 
* Main module: the Flow config file passed into `agent run MAIN_MODULE`.  
* Module loader: a component which takes a module source and runs the components defined inside.

Reply:
Given my [comment](https://github.com/grafana/agent/pull/2898#issuecomment-1414827146) about FlowModule CRDs, I think module loaders should be conceptually flexible enough that some module loaders are given a module source, while others may discover module sources on their own.

Reply:
In this RFC, I want to only focus on the 3rd use case. The other two feel like solid things to do once we have solidified the core functionality.

Reply:
Sorry, I'm confused. What use cases? I'm just suggesting we break down what we're already describing in this RFC into those three terms.

Reply:
Yes, I was suggesting that the way Flow functions now is by loading the "main module." You could say that the main module has different constraints (i.e., configuration blocks are only valid in the main module, but no submodule).